File: demos/ReactorChat/ReactorChat-Client/src/main/java/com/salesforce/servicelibs/ConsoleIterator.java
/*
 *  Copyright (c) 2017, salesforce.com, inc.
 *  All rights reserved.
 *  Licensed under the BSD 3-Clause license.
 *  For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */

package com.salesforce.servicelibs;

import jline.console.ConsoleReader;

import javax.annotation.Nonnull;
import java.io.IOException;
import java.util.Iterator;

/**
 * Adapts jLine to Iterator.
 */
public class ConsoleIterator implements Iterable<String>, Iterator<String> {
    private ConsoleReader console;
    private String prompt;
    private String lastLine;

    ConsoleIterator(ConsoleReader console, String prompt) {
        this.console = console;
        this.prompt = prompt;
    }

    @Override
    @Nonnull
    public Iterator<String> iterator() {
        return this;
    }

    @Override
    public boolean hasNext() {
        try {
            lastLine = console.readLine(prompt);
            return lastLine != null;
        } catch (IOException e) {
            return false;
        }
    }

    @Override
    public String next() {
        return lastLine;
    }
}

File: demos/reactive-grpc-chat/reactor-chat-kotlin/ReactorChat-Client-Kt/src/main/kotlin/com/salesforce/servicelibs/ConsoleIterator.kt
/*
 *  Copyright (c) 2017, salesforce.com, inc.
 *  All rights reserved.
 *  Licensed under the BSD 3-Clause license.
 *  For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */

package com.salesforce.servicelibs

import jline.console.ConsoleReader
import java.io.IOException
import javax.annotation.Nonnull

/**
 * Adapts jLine to Iterator.
 */
class ConsoleIterator internal constructor(private val console: ConsoleReader, private val prompt: String) : Iterable<String>, Iterator<String> {
    private var lastLine: String? = null

    @Nonnull
    override fun iterator(): Iterator<String> {
        return this
    }

    override fun hasNext(): Boolean {
        return try {
            lastLine = console.readLine(prompt)
            lastLine != null
        } catch (e: IOException) {
            false
        }

    }

    override fun next(): String {
        return this.lastLine!!
    }
}
